Proposed parents configuration file to be used with remap.config

jrushford opened the pull request:
This PR is to add an example parents.yaml file that eventually will be used with remap.config.  A new tag in remap say, @parent=parents.yaml may be used to configure parent/origin next hop on a per remap basis.

Looking for comments on this file.

bryancall commented on proxy/config/parents.yaml.example:
```diff
@@ -0,0 +1,60 @@
+##
+# hosts.yaml
+hosts:
+  p1: # shorthand name of host object
```
Do you need to have a shorthand name?  If not then make the host a list.

bryancall replied:
Looks like yes - Looking at primary_ring above.

zwoop commented:
We had some discussions yesterday, so this proposal will be updated from that (soon hopefully)

vmamidi commented:
Here is the updated Parent configuration.
[parent.yml.txt](https://github.com/apache/trafficserver/files/2168128/parent.yml.txt)

jrushford commented on proxy/config/parents.yaml.example:
```diff
@@ -0,0 +1,60 @@
+
+strategy:
+    protocol: h2 # allowed protocol. optional by default uses the remapped protocol
+    name: # Name of the strategy: Enum of 'consistent_hash' or 'first_live' or 'rr_strict' or 'rr_ip' or 'latched'
```
this is a list of the supported selection strategies that we know about so far.  Others can be added.  Is that what you're asking @SolidWallOfCode

jrushford commented on proxy/config/parents.yaml.example:
```diff
@@ -0,0 +1,60 @@
+      weights: # optional: weights used for the servers
+        p1: 1.0
+        p2: 2.0
+      hosts: # required list of servers
```
In our discussion we talked about defining all the hosts and their attributes in a separate file that could imported. Then in the strategy section which would be for a particular remap, you could choose the set of hosts you want for this remap and then choose how they should be weighted.

vmamidi commented:
@bryancall and @SolidWallOfCode can you guys have a look at the new file?

zwoop commented:
Yeh, I like this new format. In particular, it lets someone minimize duplications, by using anchors for shared pieces, and appropriate inheritance. Like, one could take the above one step further, and isolating the protocol specifications, healthchecks etc. as an anchored resource, and then include that. Or, one could exclude the "host" portion from those anchored sections, and override that locally in the policy sections. It's a lot nicer than all the duplications that we had initially.

SolidWallOfCode commented:
It does bring Bryan's question back - what is the use of the short name in the primary and secondary ring definitions?

SolidWallOfCode commented:
"it allows you to avoid repeating all the unchanging data about a host in every strategy file. For instance, imagine"  - but as Leif points out, YAML does this with anchors and references, as is done in the latest example configuration file. In that file, what purpose does "p1" in the "strategy" section serve? When would it be referenced instead of "*anchornameforp1"? Contrariwise, in your description, what purpose doe the "hosts" table serve? It seems to be me one is useful in the manner you describe, but I don't see why both are.

jrushford commented:
I added an '#include' to the example to show that a strategy yaml file, bar.yaml, could use '#include our hosts.yaml'. so that the two files are concatenated into one yaml document.  In the example yaml parser I've started on, I've implemented this.

SolidWallOfCode commented:
Will this be updated in accordance with the L7R Working Group meeting last week?

jrushford commented:
Alan, I’ll update that today.

From: "Alan M. Carroll" <notifications@github.com>
Reply-To: apache/trafficserver <reply@reply.github.com>
Date: Friday, August 3, 2018 at 8:08 AM
To: apache/trafficserver <trafficserver@noreply.github.com>
Cc: John Rushford <John_Rushford@cable.comcast.com>, Assign <assign@noreply.github.com>
Subject: [EXTERNAL] Re: [apache/trafficserver] Proposed parents configuration file to be used with remap.config (#3870)


Will this be updated in accordance with the L7R Working Group meeting last week?

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub<https://github.com/apache/trafficserver/pull/3870#issuecomment-410260435>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AEVybiDJBXiKWf2Nr0GoBTgf85iGTyfdks5uNFlPgaJpZM4UwC1L>.

jrushford commented:
It’s been updated, see #3870

From: "Alan M. Carroll" <notifications@github.com>
Reply-To: apache/trafficserver <reply@reply.github.com>
Date: Friday, August 3, 2018 at 8:08 AM
To: apache/trafficserver <trafficserver@noreply.github.com>
Cc: John Rushford <John_Rushford@cable.comcast.com>, Assign <assign@noreply.github.com>
Subject: [EXTERNAL] Re: [apache/trafficserver] Proposed parents configuration file to be used with remap.config (#3870)


Will this be updated in accordance with the L7R Working Group meeting last week?

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub<https://github.com/apache/trafficserver/pull/3870#issuecomment-410260435>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AEVybiDJBXiKWf2Nr0GoBTgf85iGTyfdks5uNFlPgaJpZM4UwC1L>.

vmamidi commented:
Primary and secondary group names give an impression that there are only two groups but there is no reason not to have multiple groups and transfer from group to another can happen based on failover. 
For example, 502 can result in moving to g3 whereas 404 can result in moving g2 etc..

mlibbey commented:
Sure -- just wondering how we'd configure the consistent hash feature that tells ATS to choose from group 2 if the specific chosen host in group 1 is down (as opposed to choosing another host in group 1).

jrushford commented:
good point @mlibbey, I don't see anyway to configure that so that primary/secondary order either and what should the behavior be if there are three groups?  Also, how are the multiple groups used with the round robin strategies?  Should another weight or order be added to the strategy group list as in this:

`strategy:
   policy: # Selection strategy policy: Enum of 'consistent_hash' or 'first_live' or 'rr_strict' or 'rr_ip' or 'latched'
   hash_key: # optional key to use for Hashing. Enum of 'url' or 'uri' or 'hostname' or 'path' or 'path+query' or 'cache_key' or 'path+fragment'
   groups:
     - <<: *g1
        choice: 1
     - <<: *g2
        choice: 2
`

SolidWallOfCode commented:
This is one of the reasons I suggested a recursive structure where a ring could have as elements other rings. Then it's easy - you define ring A to be a linear fail over of ring B (the primary pod) and ring C (the secondary).

I think the structure suggested in the working group, you would have ring B and ring C, and then a strategy that said "try B, then failover to C".

SolidWallOfCode commented:
@mlibbey Isn't g1 primary and g2 secondary because it's a list and g1 is first?

jrushford commented:
using the order in which the groups are in the list makes sense ie, g1 is primary because it's first.

jrushford commented:
there are two keys, in the example, "health_check" and "healthcheck".  they have different meanings where in a 'host' context it refers to a health check url and then in strategy failover context it means either an 'active' or 'passive' health check.  Is there any reason they can't be spelled the same ie, 'healthcheck' or 'health_check' to avoid confusion on which spelling is used in the differing contexts or should one be renamed entirely?  Maybe just use combine 'healthcheck' and 'url' in the host context to 'healcheck_url'?

jrushford commented:
Is the proposed format now show okay with everyone?

bryancall commented:
[approve ci fedora]

bryancall commented:
[approve ci rat]

jrushford commented:
@bryancall, @zwoop, @vmamidi, @SolidWallOfCode I don't think we need this PR any-longer and it should be closed provided everyone is okay with #5960 and the new format of strategies.yaml

bryancall commented:
@jrushford sounds good
